### DISCUSSION TITLE
My MicroDot application or coroutines silently dies after some time

### DISCUSSION THREAD

**Discussion by neliel123** [miguelgrinberg/microdot#182] [General]
Hi,
I noticed that when my ESP32 runs my MicroDot application it runs okay and continuously serves my web application as long as I continually look at my phone and never allow it to show the lock screen.

But then if my phone has been away for some time and I try to access the application again using the same IP,  I could not access my web application anymore, it just hangs and is in the loading stage on my phone browser.

For me to be able to access the application I need to hard reset my ESP32 losing some of my data in memory.

I am using the Asynchronous MicroDot and asynchronous WebSocket extension and putting my ESP32 in Access Point Mode.  Everything works fine when I am looking at my application continuously.

What could be the problem?

**Comment by neliel123** [comment]
Sorry, I made a mistake in the template engine processing.  There was some business logic that caused template errors and it caused the MicroDot application to hang.

Is there a way to handle template processing errors so that the application would get notified?  Not sure if this is already in the library though,

**Comment by neliel123** [comment]
Hey,
I have just realized that my Asynchronous MicroDot application silently dies after some time.  I could not access it anymore but it would work for hours sometimes with no issues until it died silently and I could not access my web application anymore.

Here are some of the errors that I am seeing and they are random.

```
Traceback (most recent call last):
  File "main.py", line 139, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in run_until_complete
  File "main.py", line 136, in main
  File "asyncio/funcs.py", line 1, in gather
  File "asyncio/core.py", line 1, in run_until_complete
  File "main.py", line 84, in wait_for_message
MemoryError: memory allocation failed, allocating 108 bytes

Traceback (most recent call last):
  File "main.py", line 139, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in wait_io_event
OSError: [Errno 5] EIO
```

I know that my MicroDot coroutines just died because the other coroutines are still running.  Is there a way to restart my application when an error like this comes up?
Any tips how to troubleshoot this?

**Reply by miguelgrinberg** [reply]
You have memory allocation errors. Either you are using more memory than your device has, or you have a memory leak in your application.

**Reply by neliel123** [reply]
Thanks, Miguel.  Is there a way to catch these errors and just restart my MicroControllers or do a hard reset?

**Reply by miguelgrinberg** [reply]
I assume you can catch this error in main.py line 139, which appears to be the place in your application where you start the web server?

**Reply by neliel123** [reply]
I have added the try catch already with a machine.reset on that line where I started my web server. But I never see my microcontroller getting reset.
The MicroDot coroutine just fails silently while the other coroutines still works.
The thing that bothers me is that it works fine for 1 hour or so until it stops responding..
Seems like a memory issue but I don't have an idea how to fix as it works perfectly until it stops working.

**Reply by miguelgrinberg** [reply]
> The MicroDot coroutine just fails silently while the other coroutines still works.

Not really possible. There has to be an error for the server to stop running.

> Seems like a memory issue but I don't have an idea how to fix as it works perfectly until it stops working.

Monitor memory usage to see if it grows over time. I really don't know what tools there are in MicroPython to do this, but you should look at your code and see if you might be accumulating data anywhere, which would explain your device eventually running out of memory.
